libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts

```diff
@@ -138,16 +140,29 @@ export class TeamsSSOTokenExchangeMiddleware implements Middleware {
         let tokenExchangeResponse: TokenResponse;
         const tokenExchangeRequest: TokenExchangeInvokeRequest = context.activity.value;
 
-        const tokenProvider = ExchangeToken.parse(context.adapter);
+        const tokenProvider = ExchangeToken.safeParse(context.adapter);
 
-        // TODO(jgummersall) convert to new user token client provider when available
         try {
-            tokenExchangeResponse = await tokenProvider.exchangeToken(
-                context,
-                this.oAuthConnectionName,
-                context.activity.channelId,
-                { token: tokenExchangeRequest.token }
+            const userTokenClient = context.turnState.get<UserTokenClient>(
+                (context.adapter as CloudAdapterBase).UserTokenClientKey
             );
+            if (userTokenClient) {
+                tokenExchangeResponse = await userTokenClient.exchangeToken(
+                    context.activity.from.id,
+                    this.oAuthConnectionName,
+                    context.activity.channelId,
+                    { token: tokenExchangeRequest.token }
+                );
+            } else if (tokenProvider.success) {
+                tokenExchangeResponse = await tokenProvider.data.exchangeToken(
+                    context,
+                    this.oAuthConnectionName,
+                    context.activity.channelId,
+                    { token: tokenExchangeRequest.token }
+                );
+            } else {
+                new Error('The provided token is not supported by the current adapter.');
```

Comment:
In dotnet the message is `"Token Exchange is not supported by the current adapter."`, should we maintain the same message or use this one?.
I think, the more similar to dotnet's the better.